pandas/tests/frame/methods/test_first_and_last.py

```diff
@@ -95,3 +99,48 @@ def test_empty_not_input(self, func):
         result = getattr(df, func)(offset=1)
         tm.assert_frame_equal(df, result)
         assert df is not result
+
+    @pytest.mark.parametrize("start, periods", [("2010-03-31", 1), ("2010-03-30", 2)])
+    def test_first_with_first_day_last_of_m_DO(self, frame_or_series, start, periods):
+        x = frame_or_series([1] * 100, index=bdate_range(start, periods=100))
+        result = x.first(pd.DateOffset(days=periods))
+        expected = frame_or_series(
+            [1] * periods, index=bdate_range(start, periods=periods)
+        )
+        tm.assert_equal(result, expected)
+
+    def test_first_with_first_day_end_of_frq_n_greater_one_DO(self, frame_or_series):
+        x = frame_or_series([1] * 100, index=bdate_range("2010-03-31", periods=100))
+        result = x.first(pd.DateOffset(days=2))
+        expected = frame_or_series(
+            [1] * 2, index=bdate_range("2010-03-31", "2010-04-01")
+        )
+        tm.assert_equal(result, expected)
+
+    def test_first_w_DateOffset(self):
+        # GH#51284
+        i = pd.date_range("2018-04-09", periods=4, freq="2D")
+        x = DataFrame({"A": [1, 2, 3, 4]}, index=i)
+        result = x.first(pd.DateOffset(days=3))
+        expected = DataFrame(
+            {"A": [1, 2]}, index=pd.date_range("2018-04-09", periods=2, freq="2D")
+        )
```

Comment:
for this one, perhaps we could construct the output explicitly? something like
```python
expected = pd.DataFrame({'A':[1,2]}, index=to_datetime(['2018-04-09', '2018-04-11']))
```
?

just because `date_range` also uses offsets, and this would guard against potential bugs in that changing the test (and also because this is a regression test linked to a specific issue, so it's probably best to stay as simple and explicit as possible)

Reply:
Thanks for correcting and explaining

Reply:
I'm also using `date_range` on all the tests I wrote, I'm thinking I should remove it on all my tests..?
If I avoid `date_range` on `expected`, it seems that I have to eliminate it from the input too, as `tm.assert_equal` fails. Or I have to add the `freq`. Does this look horrible?:
```
expected = (
            pd.DataFrame({"A":[1, 2]},
            index=pd.to_datetime(['2018-04-09', '2018-04-11']))
            ).resample('2D').ffill()
```

Reply:
Maybe using `resample` is wrong..  I should just write by hand the whole.
Sorry for the noise